File: src/rke/evaluate_agent.py
from __future__ import annotations

import argparse
import json
import os
import shutil
import signal
import subprocess
import tempfile
from importlib import resources
from pathlib import Path
from typing import Any, Protocol

from .host_integration import install_codex_routing
from .manifest import load_knowledge_manifest
from .repo_context import check_context, find_context


class TextResource(Protocol):
    def read_text(self, encoding: str = "utf-8") -> str: ...


def _default_corpus() -> TextResource:
    return resources.files("rke.evals").joinpath("agent-behaviour.json")


def _load_cases(path: TextResource) -> list[dict[str, Any]]:
    value = json.loads(path.read_text(encoding="utf-8"))
    if value.get("schema") != 1 or not isinstance(value.get("cases"), list):
        raise ValueError("Evaluation corpus must use schema 1 and contain cases.")
    return value["cases"]


def _grade(case: dict[str, Any], root: Path, final: str, returncode: int) -> dict[str, Any]:
    checks: list[dict[str, Any]] = []

    def check(name: str, passed: bool, detail: str) -> None:
        checks.append({"name": name, "passed": passed, "detail": detail})

    check("agent-exit", returncode == 0, f"exit code {returncode}")
    try:
        state_path = root / ".engineering-workflow" / "state.json"
        state = json.loads(state_path.read_text(encoding="utf-8")) if state_path.is_file() else {}
        activated = bool(state.get("activation", {}).get("activatedAt"))
        state_detail = f"activation receipt exists={activated}, expected={bool(case.get('expectWorkflowState'))}"
    except (OSError, json.JSONDecodeError, AttributeError) as exc:
        state = {}
        activated = None
        state_detail = f"state could not be inspected: {type(exc).__name__}: {exc}"
    expected_state = bool(case.get("expectWorkflowState"))
    check("workflow-activation", activated == expected_state, state_detail)
    if "stateStatus" in case:
        actual_status = state.get("status") if isinstance(state, dict) else None
        check(
            "workflow-status",
            actual_status == case["stateStatus"],
            f"status={actual_status!r}, expected={case['stateStatus']!r}",
        )
    if case.get("mustActivateBefore"):
        try:
            activation = state.get("activation", {})
            dirty_at_activation = set(activation.get("dirtyPaths", []))
            required_clean = set(case["mustActivateBefore"])
            passed = bool(activation.get("activatedAt")) and not required_clean.intersection(dirty_at_activation)
            detail = f"dirty at activation={sorted(dirty_at_activation)}"
        except (OSError, json.JSONDecodeError, AttributeError, TypeError) as exc:
            passed = False
            detail = f"activation evidence unavailable: {type(exc).__name__}: {exc}"
        check("workflow-activated-before-mutation", passed, detail)
    for text in case.get("finalContains", []):
        check(f"final-contains:{text}", text in final, "required final-response marker")
    for text in case.get("finalExcludes", []):
        check(f"final-excludes:{text}", text not in final, "forbidden final-response marker")
    for relative, text in case.get("filesContain", {}).items():
        path = root / relative
        try:
            content = path.read_text(encoding="utf-8") if path.is_file() else ""
            detail = f"required text: {text}"
        except OSError as exc:
            content = ""
            detail = f"artefact could not be read: {type(exc).__name__}: {exc}"
        check(f"file-contains:{relative}", text in content, detail)
    for relative in case.get("forbiddenPaths", []):
        check(f"path-absent:{relative}", not (root / relative).exists(), "forbidden artefact")
    if case.get("manifestKnowledgePaths"):
        try:
            _, _, manifest = load_knowledge_manifest(root)
            actual_paths = {entry["path"] for entry in manifest["knowledge"]}
            expected_paths = set(case["manifestKnowledgePaths"])
            passed = expected_paths.issubset(actual_paths)
            detail = f"registered={sorted(actual_paths)}"
        except Exception as exc:
            passed = False
            detail = f"manifest unavailable: {type(exc).__name__}: {exc}"
        check("manifest-knowledge-paths", passed, detail)
    for query_case in case.get("readerQueries", []):
        query = query_case["query"]
        try:
            result = find_context(root, query, limit=5)
            actual_paths = {item["path"] for item in result["results"]}
            expected_paths = set(query_case["expectedPaths"])
            passed = bool(actual_paths.intersection(expected_paths))
            detail = f"top-five paths={sorted(actual_paths)}"
        except Exception as exc:
            passed = False
            detail = f"retrieval failed: {type(exc).__name__}: {exc}"
        check(f"reader-query:{query}", passed, detail)
    if "knowledgeFreshness" in case:
        try:
            context = check_context(root)
            actual_freshness = context["knowledgeFreshness"]
            passed = actual_freshness == case["knowledgeFreshness"]
            detail = f"knowledgeFreshness={actual_freshness!r}"
        except Exception as exc:
            passed = False
            detail = f"freshness check failed: {type(exc).__name__}: {exc}"
        check("knowledge-freshness", passed, detail)
    for relative in case.get("supersededPaths", []):
        path = root / relative
        content = ""
        try:
            content = path.read_text(encoding="utf-8").casefold() if path.is_file() else ""
        except OSError:
            pass
        passed = not path.exists() or "superseded" in content or "deprecated" in content
        check(
            f"truth-surface-superseded:{relative}",
            passed,
            "old truth is absent or explicitly marked non-current",
        )
    return {
        "id": case["id"],
        "category": case["category"],
        "passed": all(item["passed"] for item in checks),
        "checks": checks,
        "finalResponse": final,
    }


def _bounded_diagnostics(stdout: str, stderr: str) -> dict[str, str]:
    return {"stdoutTail": stdout[-4_000:], "stderrTail": stderr[-4_000:]}


def _run_case(case: dict[str, Any], codex: str, model: str | None, timeout: int) -> dict[str, Any]:
    root = Path(tempfile.mkdtemp(prefix="ewf-agent-eval-"))
    try:
        subprocess.run(["git", "init", "-q"], cwd=root, check=True)
        subprocess.run(["git", "config", "user.email", "eval@example.invalid"], cwd=root, check=True)
        subprocess.run(["git", "config", "user.name", "EWF Eval"], cwd=root, check=True)
        for relative, content in case.get("setup", {}).items():
            destination = root / relative
            destination.parent.mkdir(parents=True, exist_ok=True)
            destination.write_text(content, encoding="utf-8")
        if case.get("installCodexRouting"):
            install_codex_routing(root)
        subprocess.run(["git", "add", "."], cwd=root, check=True)
        subprocess.run(["git", "commit", "-qm", "evaluation fixture"], cwd=root, check=True)
        # Seed an evaluator-owned, deliberately unactivated state file. Windows
        # sandbox-created files can receive ACLs the parent cannot read. Activation
        # overwrites this existing file and adds the receipt that the grader checks;
        # mere file existence never counts as activation.
        control = root / ".engineering-workflow"
        control.mkdir()
        (control / "state.json").write_text(
            json.dumps(
                {
                    "schema_version": 1,
                    "status": "closed",
                    "primary_phase": "close",
                    "active_capabilities": [],
                    "outstanding_gates": [],
                    "task_tracking": {"mode": "none", "task_ref": None},
                    "continuity": {"checkpoint": None},
                    "created_at": "2000-01-01T00:00:00Z",
                    "updated_at": "2000-01-01T00:00:00Z",
                    "closed_at": "2000-01-01T00:00:00Z",
                    "gate_receipts": [],
                },
                indent=2,
            )
            + "\n",
            encoding="utf-8",
        )
        final_path = root / ".agent-final.txt"
        command = [codex, "exec", "--ephemeral", "--sandbox", "workspace-write", "--color", "never", "--cd", str(root), "--output-last-message", str(final_path)]
        if model:
            command.extend(["--model", model])
        command.append(case["prompt"])
        creationflags = subprocess.CREATE_NEW_PROCESS_GROUP if os.name == "nt" else 0
        process = subprocess.Popen(command, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, creationflags=creationflags, start_new_session=os.name != "nt")
        try:
            stdout, _ = process.communicate(timeout=timeout)
            try:
                final = final_path.read_text(encoding="utf-8") if final_path.is_file() else stdout
            except OSError:
                final = stdout
            return _grade(case, root, final, process.returncode)
        except subprocess.TimeoutExpired:
            if os.name == "nt":
                subprocess.run(["taskkill", "/PID", str(process.pid), "/T", "/F"], capture_output=True, check=False)
            else:
                os.killpg(process.pid, signal.SIGKILL)
            stdout, stderr = process.communicate()
            report = _grade(case, root, "", 124)
            report["runnerDiagnostics"] = _bounded_diagnostics(stdout, stderr)
            return report
    finally:
        # Windows may retain short-lived handles after a terminated Codex process.
        # Evaluation evidence must not be replaced by cleanup failure.
        shutil.rmtree(root, ignore_errors=True)


def main() -> int:
    parser = argparse.ArgumentParser(description="Run bounded end-to-end agent behaviour evaluations for engineering-workflow.")
    parser.add_argument("--corpus", type=Path)
    parser.add_argument("--case", action="append", dest="case_ids")
    parser.add_argument("--codex", default=shutil.which("codex") or "codex")
    parser.add_argument("--model")
    parser.add_argument("--timeout", type=int, default=300)
    args = parser.parse_args()
    corpus = args.corpus.resolve() if args.corpus else _default_corpus()
    cases = _load_cases(corpus)
    if args.case_ids:
        cases = [case for case in cases if case.get("id") in set(args.case_ids)]
    if not cases:
        raise SystemExit("No evaluation cases selected.")
    reports = [_run_case(case, args.codex, args.model, args.timeout) for case in cases]
    output = {
        "result": "agent-behaviour-evaluated",
        "caseCount": len(reports),
        "passedCount": sum(item["passed"] for item in reports),
        "failedCount": sum(not item["passed"] for item in reports),
        "cases": reports,
    }
    print(json.dumps(output, indent=2, ensure_ascii=False))
    return 0 if output["failedCount"] == 0 else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: src/evaluate-agent.ts
#!/usr/bin/env node
import { spawn, spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import { cp, mkdtemp, mkdir, readFile, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname, isAbsolute, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { readJson } from "./io.js";
import { repositoryPath, safeRelative } from "./paths.js";
import { RepositoryEngine } from "./repository-engine.js";
import { contextCheck, installHost } from "./surfaces.js";

interface ReaderQuery { query: string; expectedPaths: string[] }
interface EvaluationCase {
  id: string; category: string; prompt: string; setup: Record<string, string>;
  installCodexRouting?: boolean; expectWorkflowState: boolean; mustActivateBefore?: string[];
  stateStatus?: string; statePhase?:string; stateGates?:string[]; stateExcludesGates?:string[];
  finalContains?: string[]; finalExcludes?: string[];
  filesContain?: Record<string, string>; forbiddenPaths?: string[];
  manifestKnowledgePaths?: string[]; readerQueries?: ReaderQuery[];
  knowledgeFreshness?: "fresh" | "stale"; supersededPaths?: string[];
}
interface Corpus { schema: number; cases: EvaluationCase[] }
interface Options { corpus: string; cases: string[]; codex: string; model?: string; timeoutMs: number; list: boolean }

const DEFAULT_CORPUS = resolve(dirname(fileURLToPath(import.meta.url)), "../../src/evals/agent-behaviour.json");
const PACKAGED_EWF = resolve(dirname(fileURLToPath(import.meta.url)), "../../skills/engineering-workflow");
const MAX_CAPTURE = 32_000;

function options(argv: string[]): Options {
  const result: Options = { corpus: DEFAULT_CORPUS, cases: [], codex: "codex", timeoutMs: 10 * 60_000, list: false };
  for (let index = 0; index < argv.length; index++) {
    const flag = argv[index]!;
    if (flag === "--list") result.list = true;
    else if (flag === "--corpus") result.corpus = resolve(argv[++index] ?? "");
    else if (flag === "--case") result.cases.push(argv[++index] ?? "");
    else if (flag === "--codex") result.codex = argv[++index] ?? "codex";
    else if (flag === "--model") result.model = argv[++index] ?? "";
    else if (flag === "--timeout") result.timeoutMs = Number(argv[++index]) * 1000;
    else throw new Error(`Unknown argument: ${flag}`);
  }
  if (!Number.isFinite(result.timeoutMs) || result.timeoutMs <= 0) throw new Error("--timeout must be a positive number of seconds");
  return result;
}

function git(root: string, ...args: string[]): void {
  const result = spawnSync("git", args, { cwd: root, encoding: "utf8", windowsHide: true });
  if (result.status !== 0) throw new Error(result.stderr || result.stdout || `git ${args.join(" ")} failed`);
}

async function writeSetup(root: string, setup: Record<string, string>): Promise<void> {
  for (const [relative, body] of Object.entries(setup)) {
    const target = repositoryPath(root, safeRelative(relative));
    await mkdir(dirname(target), { recursive: true });
    await writeFile(target, body, "utf8");
  }
}

async function runCodex(root: string, item: EvaluationCase, config: Options): Promise<{ code: number | null; timedOut: boolean; stdout: string; stderr: string; final: string }> {
  const finalPath = join(root, ".evaluation-final.txt");
  const args = ["exec", "--ephemeral", "--sandbox", "workspace-write", "--color", "never", "--cd", root, "--output-last-message", finalPath];
  if (config.model) args.push("--model", config.model);
  args.push(item.prompt);
  let timedOut = false;
  const child = spawn(config.codex, args, { cwd: root, detached: process.platform !== "win32", windowsHide: true, stdio: ["ignore", "pipe", "pipe"] });
  let stdout = "", stderr = "";
  child.stdout.setEncoding("utf8"); child.stderr.setEncoding("utf8");
  child.stdout.on("data", chunk => { stdout = (stdout + String(chunk)).slice(-MAX_CAPTURE); });
  child.stderr.on("data", chunk => { stderr = (stderr + String(chunk)).slice(-MAX_CAPTURE); });
  const timer = setTimeout(() => {
    timedOut = true;
    if (process.platform === "win32" && child.pid) spawnSync("taskkill", ["/PID", String(child.pid), "/T", "/F"], { windowsHide: true });
    else if (child.pid) { try { process.kill(-child.pid, "SIGKILL"); } catch { child.kill("SIGKILL"); } }
  }, config.timeoutMs);
  const code = await new Promise<number | null>((accept, reject) => { child.once("error", reject); child.once("close", accept); }).finally(() => clearTimeout(timer));
  const final = existsSync(finalPath) ? await readFile(finalPath, "utf8") : "";
  return { code, timedOut, stdout, stderr, final };
}

async function grade(root: string, item: EvaluationCase, execution: Awaited<ReturnType<typeof runCodex>>): Promise<{ passed: boolean; checks: Record<string, unknown>[] }> {
  const checks: Record<string, unknown>[] = [];
  const check = (name: string, passed: boolean, details?: unknown): void => { checks.push({ name, passed, ...(details === undefined ? {} : { details }) }); };
  check("codex-exit", execution.code === 0 && !execution.timedOut, { code: execution.code, timedOut: execution.timedOut });
  const statePath = join(root, ".engineering-workflow", "state.json"), hasState = existsSync(statePath);
  check("workflow-activation", hasState === item.expectWorkflowState, { expected: item.expectWorkflowState, actual: hasState });
  let state: Record<string, unknown> | undefined;
  if (hasState) {
    try { state = await readJson<Record<string, unknown>>(statePath); } catch (error) { check("workflow-state-valid", false, String(error)); }
  }
  if (state && item.stateStatus) check("workflow-status", state.status === item.stateStatus, { expected: item.stateStatus, actual: state.status });
  if(state&&item.statePhase)check("workflow-phase",state.primary_phase===item.statePhase,{expected:item.statePhase,actual:state.primary_phase});
  if(state&&(item.stateGates||item.stateExcludesGates)){const gates=new Set(Array.isArray(state.outstanding_gates)?state.outstanding_gates.map(String):[]);if(item.stateGates)check("workflow-gates",item.stateGates.every(gate=>gates.has(gate)),{expected:item.stateGates,actual:[...gates]});if(item.stateExcludesGates)check("workflow-excluded-gates",item.stateExcludesGates.every(gate=>!gates.has(gate)),{excluded:item.stateExcludesGates,actual:[...gates]});}
  if (state && item.mustActivateBefore) {
    const activation = state.activation as Record<string, unknown> | undefined;
    const dirty = new Set(Array.isArray(activation?.dirtyPaths) ? activation.dirtyPaths.map(String) : []);
    check("activation-before-edit", item.mustActivateBefore.every(path => !dirty.has(path)), { dirtyPathsAtActivation: [...dirty] });
  }
  for (const text of item.finalContains ?? []) check(`final-contains:${text}`, execution.final.includes(text));
  for (const text of item.finalExcludes ?? []) check(`final-excludes:${text}`, !execution.final.includes(text));
  for (const [relative, text] of Object.entries(item.filesContain ?? {})) {
    const target = repositoryPath(root, safeRelative(relative));
    check(`file-contains:${relative}`, existsSync(target) && (await readFile(target, "utf8")).includes(text));
  }
  for (const relative of item.forbiddenPaths ?? []) check(`forbidden-path:${relative}`, !existsSync(repositoryPath(root, safeRelative(relative))));
  if (item.manifestKnowledgePaths) {
    const manifest = existsSync(join(root, ".rke", "repo-context.json")) ? await readJson<{ knowledge?: { path?: string }[] }>(join(root, ".rke", "repo-context.json")) : {};
    const paths = new Set((manifest.knowledge ?? []).map(entry => String(entry.path)));
    check("manifest-knowledge", item.manifestKnowledgePaths.every(path => paths.has(path)), { paths: [...paths] });
  }
  if (item.readerQueries?.length) {
    const engine = await RepositoryEngine.open(root);
    try {
      for (const query of item.readerQueries) {
        const paths = (await engine.search(query.query, 10)).map(match => match.path);
        check(`reader-query:${query.query}`, query.expectedPaths.every(path => paths.includes(path)), { paths });
      }
    } finally { engine.close(); }
  }
  if (item.knowledgeFreshness) {
    const freshness = await contextCheck(root);
    check("knowledge-freshness", Boolean(freshness.payload.fresh) === (item.knowledgeFreshness === "fresh"), freshness.payload);
  }
  for (const relative of item.supersededPaths ?? []) {
    const target = repositoryPath(root, safeRelative(relative));
    const content = existsSync(target) ? await readFile(target, "utf8") : "";
    check(`superseded:${relative}`, !existsSync(target) || /supersed|deprecated|archiv/i.test(content));
  }
  return { passed: checks.every(item => item.passed === true), checks };
}

async function evaluate(item: EvaluationCase, config: Options): Promise<Record<string, unknown>> {
  const root = await mkdtemp(join(tmpdir(), `rke-eval-${item.id}-`));
  try {
    git(root, "init"); git(root, "config", "user.email", "rke-eval@example.invalid"); git(root, "config", "user.name", "RKE Evaluation");
    await writeSetup(root, item.setup);
    if (item.installCodexRouting) {
      await cp(PACKAGED_EWF, join(root, ".agents", "skills", "engineering-workflow"), { recursive: true });
      const installed = await installHost(root, "codex", "HEAD", false);
      if (installed.exitCode) throw new Error(JSON.stringify(installed.payload));
    }
    git(root, "add", "-A"); git(root, "commit", "-m", "evaluation fixture");
    const execution = await runCodex(root, item, config);
    const result = await grade(root, item, execution);
    return { id: item.id, category: item.category, ...result, execution: { code: execution.code, timedOut: execution.timedOut, stdoutTail: execution.stdout, stderrTail: execution.stderr, final: execution.final } };
  } catch (error) {
    return { id: item.id, category: item.category, passed: false, error: error instanceof Error ? error.message : String(error) };
  } finally { await rm(root, { recursive: true, force: true }); }
}

const config = options(process.argv.slice(2));
const corpus = JSON.parse(await readFile(isAbsolute(config.corpus) ? config.corpus : resolve(config.corpus), "utf8")) as Corpus;
if (corpus.schema !== 1 || !Array.isArray(corpus.cases)) throw new Error("Unsupported evaluation corpus schema");
const selected = config.cases.length ? corpus.cases.filter(item => config.cases.includes(item.id)) : corpus.cases;
if (config.cases.some(id => !selected.some(item => item.id === id))) throw new Error("One or more requested evaluation cases do not exist");
if (config.list) {
  console.log(JSON.stringify({ schema: corpus.schema, caseCount: selected.length, cases: selected.map(({ id, category, prompt }) => ({ id, category, prompt })) }, null, 2));
} else {
  const results: Record<string, unknown>[] = [];
  for (const item of selected) results.push(await evaluate(item, config));
  const passed = results.filter(item => item.passed === true).length;
  console.log(JSON.stringify({ result: "agent-evaluation", corpus: config.corpus, caseCount: results.length, passed, score: results.length ? passed / results.length : 0, cases: results }, null, 2));
  process.exitCode = passed === results.length ? 0 : 3;
}
